File: flink/datastream/src/main/java/com/benchmark/IcebergIngestJob.java
package com.benchmark;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.data.GenericRowData;
import org.apache.flink.table.data.RowData;
import org.apache.flink.table.data.StringData;
import org.apache.iceberg.DistributionMode;
import org.apache.iceberg.PartitionSpec;
import org.apache.iceberg.Schema;
import org.apache.iceberg.catalog.Namespace;
import org.apache.iceberg.catalog.TableIdentifier;
import org.apache.iceberg.flink.CatalogLoader;
import org.apache.iceberg.flink.TableLoader;
import org.apache.iceberg.flink.sink.IcebergSink;
import org.apache.iceberg.types.Types;

import java.util.HashMap;
import java.util.Map;

/**
 * Flink DataStream → Iceberg ingest benchmark job.
 *
 * <p>Kafka (JSON) → parse → RowData → {@link IcebergSink} (SinkV2). The write
 * distribution mode, format version, and upsert behaviour are all controlled by
 * job args so a single jar drives every feature-comparison run.
 *
 * <p>Args (all optional, sensible defaults from env):
 * <pre>
 *   --bootstrap  kafka:9092
 *   --topic      events
 *   --catalog-uri http://iceberg-rest:8181
 *   --warehouse  s3://warehouse/
 *   --database   streaming
 *   --table      events
 *   --distribution-mode none|hash|range   (default hash)
 *   --upsert     true|false               (default false)
 *   --parallelism N
 * </pre>
 */
public final class IcebergIngestJob {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    // Iceberg schema mirrored from common/event_schema.md, plus event_time: a
    // real timestamp column so we can partition by hours(event_time) (the
    // timestamp-partitioning axis) as an alternative to bucket(user_id).
    private static final Schema SCHEMA = new Schema(
            Types.NestedField.required(1, "event_id", Types.StringType.get()),
            Types.NestedField.required(2, "user_id", Types.LongType.get()),
            Types.NestedField.optional(3, "event_type", Types.StringType.get()),
            Types.NestedField.optional(4, "product_id", Types.LongType.get()),
            Types.NestedField.optional(5, "country", Types.StringType.get()),
            Types.NestedField.optional(6, "amount", Types.DoubleType.get()),
            Types.NestedField.optional(7, "currency", Types.StringType.get()),
            Types.NestedField.optional(8, "quantity", Types.IntegerType.get()),
            Types.NestedField.optional(9, "event_ts", Types.LongType.get()),
            Types.NestedField.optional(10, "ingest_ts", Types.LongType.get()),
            Types.NestedField.optional(11, "payload", Types.StringType.get()),
            Types.NestedField.optional(12, "event_time", Types.TimestampType.withoutZone()),
            // epoch-minute bucket for minute-granular time partitioning (Iceberg
            // has no native minute() transform, so we identity-partition on this).
            Types.NestedField.optional(13, "event_minute", Types.LongType.get()));

    public static void main(String[] args) throws Exception {
        ParamUtil p = ParamUtil.fromArgs(args);
        String bootstrap = p.get("bootstrap", System.getenv().getOrDefault("KAFKA_BOOTSTRAP", "kafka:9092"));
        String topic = p.get("topic", System.getenv().getOrDefault("SOURCE_TOPIC", "events"));
        String catalogUri = p.get("catalog-uri", System.getenv().getOrDefault("CATALOG_URI", "http://iceberg-rest:8181"));
        String warehouse = p.get("warehouse", "s3://" + System.getenv().getOrDefault("WAREHOUSE_BUCKET", "warehouse") + "/");
        String database = p.get("database", System.getenv().getOrDefault("ICEBERG_DB", "streaming"));
        String table = p.get("table", System.getenv().getOrDefault("ICEBERG_TABLE", "events"));
        DistributionMode mode = DistributionMode.fromName(
                p.get("distribution-mode", System.getenv().getOrDefault("WRITE_DISTRIBUTION_MODE", "hash")));
        boolean upsert = Boolean.parseBoolean(p.get("upsert", "false"));
        int parallelism = Integer.parseInt(p.get("parallelism", "4"));
        // partitioning axis: "bucket" = bucket(N,user_id) | "time" = hours(event_time)
        String partitioning = p.get("partitioning", System.getenv().getOrDefault("PARTITIONING", "bucket"));
        // bucket count = number of concurrent write streams the sink can run. The
        // upsert write-path (per-bucket data + equality-delete files) is I/O-bound, so
        // this — not core count — is the real throughput lever. Default 48 lets all 12
        // slots write in parallel (was 16, which capped concurrency below parallelism).
        int buckets = Integer.parseInt(p.get("buckets", System.getenv().getOrDefault("BUCKETS", "48")));
        String formatVersion = p.get("format-version", System.getenv().getOrDefault("TABLE_FORMAT_VERSION", "2"));

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(parallelism);
        // Commit interval is the latency floor for an Iceberg sink — make it explicit.
        long ckpt = Long.parseLong(System.getenv().getOrDefault("CHECKPOINT_INTERVAL_MS", "30000"));
        env.enableCheckpointing(ckpt);

        // earliest = chew any pre-filled backlog (overload test); latest = live only.
        String startOffsets = p.get("starting-offsets", System.getenv().getOrDefault("STARTING_OFFSETS", "latest"));
        OffsetsInitializer initial = "earliest".equalsIgnoreCase(startOffsets)
                ? OffsetsInitializer.earliest() : OffsetsInitializer.latest();
        KafkaSource<String> source = KafkaSource.<String>builder()
                .setBootstrapServers(bootstrap)
                .setTopics(topic)
                .setGroupId("flink-iceberg-bench")
                .setStartingOffsets(initial)
                .setValueOnlyDeserializer(new SimpleStringSchema())
                .build();

        DataStream<String> raw = env.fromSource(source, WatermarkStrategy.noWatermarks(), "kafka-source");
        DataStream<RowData> rows = raw.map(IcebergIngestJob::parse).name("json->rowdata");

        CatalogLoader catalogLoader = restCatalog(catalogUri, warehouse);
        TableIdentifier id = TableIdentifier.of(Namespace.of(database), table);
        ensureTable(catalogLoader, id, mode, partitioning, formatVersion, upsert, buckets);
        TableLoader tableLoader = TableLoader.fromCatalog(catalogLoader, id);

        IcebergSink.Builder builder = IcebergSink.forRowData(rows)
                .tableLoader(tableLoader)
                .distributionMode(mode)
                .writeParallelism(parallelism);
        if (upsert) {
            builder.upsert(true).equalityFieldColumns(java.util.List.of("user_id"));
        }
        builder.append();

        // INLINE table maintenance — embedded in THIS ingest job's env (single
        // env.execute), matching the AWS sample DataStreamIcebergJob. Compaction runs
        // in the same job graph, so it SHARES the ingest job's slots (the fair contrast
        // to Spark's separate spark-submit). forTable(env, tableLoader) with NO lock
        // factory → Iceberg's Flink COORDINATOR lock (state-based, PR#15151) — no
        // JDBC/ZooKeeper needed. Gated by FLINK_INLINE_MAINT=1 so we can A/B against
        // the no-maintenance baseline.
        if ("1".equals(System.getenv().getOrDefault("FLINK_INLINE_MAINT", "0"))) {
            setupInlineMaintenance(env, tableLoader);
        }

        env.execute("flink-datastream-iceberg-ingest[" + mode + (upsert ? ",upsert" : "")
                + (System.getenv().getOrDefault("FLINK_INLINE_MAINT", "0").equals("1") ? ",inline-maint" : "") + "]");
    }

    /**
     * Attach Iceberg TableMaintenance operators to the ingest env (shared resources).
     * Triggers are commit/file-count based (per AWS sample) so a streaming table that
     * accrues delete-file debt gets continuously compacted in-job.
     */
    private static void setupInlineMaintenance(StreamExecutionEnvironment env, TableLoader tableLoader) throws Exception {
        int rewriteFileCount = Integer.parseInt(System.getenv().getOrDefault("MAINT_REWRITE_FILE_COUNT", "20"));
        int expireCommits = Integer.parseInt(System.getenv().getOrDefault("MAINT_EXPIRE_COMMITS", "10"));
        int retainLast = Integer.parseInt(System.getenv().getOrDefault("MAINT_RETAIN_LAST", "5"));
        // forTable(env, tableLoader) — 2-arg overload → coordinator lock (no external lock).
        org.apache.iceberg.flink.maintenance.api.TableMaintenance
                .forTable(env, tableLoader)
                .uidSuffix("sfi-inline-maint")
                .rateLimit(java.time.Duration.ofSeconds(10))
                .add(org.apache.iceberg.flink.maintenance.api.RewriteDataFiles.builder()
                        .scheduleOnDataFileCount(rewriteFileCount)
                        .scheduleOnPosDeleteRecordCount(500_000)
                        .deleteFileThreshold(1)
                        .targetFileSizeBytes(128L * 1024 * 1024)
                        .partialProgressEnabled(true)
                        .partialProgressMaxCommits(5))
                .add(org.apache.iceberg.flink.maintenance.api.ExpireSnapshots.builder()
                        .scheduleOnCommitCount(expireCommits)
                        .maxSnapshotAge(java.time.Duration.ofMinutes(30))
                        .retainLast(retainLast)
                        .cleanExpiredMetadata(true))
                .append();
    }

    private static RowData parse(String json) throws Exception {
        JsonNode n = MAPPER.readTree(json);
        GenericRowData r = new GenericRowData(13);
        long eventTs = n.path("event_ts").asLong();
        r.setField(0, StringData.fromString(n.path("event_id").asText("")));
        r.setField(1, n.path("user_id").asLong());
        r.setField(2, StringData.fromString(n.path("event_type").asText("")));
        r.setField(3, n.path("product_id").asLong());
        r.setField(4, StringData.fromString(n.path("country").asText("")));
        r.setField(5, n.path("amount").asDouble());
        r.setField(6, StringData.fromString(n.path("currency").asText("")));
        r.setField(7, n.path("quantity").asInt());
        r.setField(8, eventTs);
        r.setField(9, n.path("ingest_ts").asLong());
        r.setField(10, StringData.fromString(n.path("payload").asText("")));
        r.setField(11, org.apache.flink.table.data.TimestampData.fromEpochMillis(eventTs));
        r.setField(12, eventTs / 60_000L); // epoch minute
        return r;
    }

    /**
     * Local runs use the Iceberg REST catalog on MinIO; EKS runs set
     * CATALOG_TYPE=glue (or pass --catalog-uri glue) to use AWS Glue + real S3.
     * The rest of the job is identical — this is the only local↔cloud binding.
     */
    static CatalogLoader restCatalog(String uri, String warehouse) {  // package-private: reused by FlinkMaintenanceJob
        String catalogType = System.getenv().getOrDefault("CATALOG_TYPE", "rest");
        org.apache.hadoop.conf.Configuration conf = new org.apache.hadoop.conf.Configuration();

        if ("glue".equalsIgnoreCase(catalogType) || "glue".equalsIgnoreCase(uri)) {
            String region = System.getenv().getOrDefault("AWS_REGION", "us-east-1");
            Map<String, String> props = new HashMap<>();
            props.put("warehouse", warehouse);
            props.put("io-impl", "org.apache.iceberg.aws.s3.S3FileIO");
            props.put("client.region", region);
            props.put("s3.region", region);
            // Explicit REGIONAL S3 endpoint: without it the SDK uses virtual-hosted
            // global "bucket.s3.amazonaws.com", whose wildcard cert doesn't cover a
            // multi-dot bucket name → SSLPeerUnverifiedException. The regional host
            // "s3.<region>.amazonaws.com" is covered by "*.s3.<region>.amazonaws.com".
            props.put("s3.endpoint", "https://s3." + region + ".amazonaws.com");
            return CatalogLoader.custom(
                    "demo", props, conf, "org.apache.iceberg.aws.glue.GlueCatalog");
        }

        Map<String, String> props = new HashMap<>();
        props.put("uri", uri);
        props.put("warehouse", warehouse);
        props.put("io-impl", "org.apache.iceberg.aws.s3.S3FileIO");
        props.put("s3.endpoint", System.getenv().getOrDefault("S3_ENDPOINT", "http://minio:9000"));
        props.put("s3.path-style-access", "true");
        return CatalogLoader.rest("demo", conf, props);
    }

    private static void ensureTable(CatalogLoader loader, TableIdentifier id, DistributionMode mode,
                                    String partitioning, String formatVersion, boolean upsert, int buckets) {
        // Works for any Catalog impl (REST locally, Glue on EKS).
        org.apache.iceberg.catalog.Catalog catalog = loader.loadCatalog();
        try {
            // The REST/Glue catalog requires the namespace to exist before createTable.
            if (catalog instanceof org.apache.iceberg.catalog.SupportsNamespaces) {
                org.apache.iceberg.catalog.SupportsNamespaces ns =
                        (org.apache.iceberg.catalog.SupportsNamespaces) catalog;
                if (!ns.namespaceExists(id.namespace())) {
                    ns.createNamespace(id.namespace());
                }
            }
            // Clean per-run state: drop+purge the table so each run measures only
            // its own output (RECREATE_TABLE=1). purge=true removes the data files
            // too — the engine self-cleans, no gated AWS CLI call needed.
            if ("1".equals(System.getenv().getOrDefault("RECREATE_TABLE", "0")) && catalog.tableExists(id)) {
                catalog.dropTable(id, true);
            }
            if (!catalog.tableExists(id)) {
                // Partitioning axis: by cardinality (bucket on the upsert key) or by time.
                // "time" = identity on epoch-minute (≈5 partitions in a 5-min run,
                // so time-partitioning fan-out is actually exercised); "bucket" =
                // 16 hash buckets on the upsert key (cardinality partitioning).
                PartitionSpec spec = "time".equalsIgnoreCase(partitioning)
                        ? PartitionSpec.builderFor(SCHEMA).identity("event_minute").build()
                        : PartitionSpec.builderFor(SCHEMA).bucket("user_id", buckets).build();  // N concurrent write streams
                Map<String, String> props = new HashMap<>();
                props.put("format-version", formatVersion);
                props.put("write.distribution-mode", mode.modeName());
                if (upsert) {
                    props.put("write.upsert.enabled", "true");
                }
                catalog.createTable(id, SCHEMA, spec, props);
            }
        } catch (Exception e) {
            throw new RuntimeException("failed to ensure table " + id, e);
        } finally {
            if (catalog instanceof AutoCloseable) {
                try { ((AutoCloseable) catalog).close(); } catch (Exception ignore) { /* best effort */ }
            }
        }
    }
}

File: spark/scala/src/main/scala/com/benchmark/IcebergIngestJob.scala
package com.benchmark

import java.util.concurrent.{Executors, Semaphore}

import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions._
import org.apache.spark.sql.streaming.Trigger
import org.apache.spark.sql.types._

/**
 * Spark Structured Streaming → Iceberg ingest benchmark.
 *
 * Kafka (JSON) → parse → Iceberg table via the DataSourceV2 streaming writer.
 * Micro-batch with a tight processingTime trigger; the trigger interval and the
 * Iceberg commit boundary coincide, so — as with Flink checkpoints — the commit
 * interval is the end-to-end latency floor for the table sink.
 *
 * Write distribution mode is a *table property* on the Iceberg side; Spark
 * additionally honours `write.distribution-mode` when planning the write, and
 * `fanout-enabled` controls whether partitions are written without a pre-sort.
 *
 * Args: --distribution-mode none|hash|range  --upsert true|false  --trigger 30
 */
object IcebergIngestJob {

  private val schema = StructType(Seq(
    StructField("event_id", StringType),
    StructField("user_id", LongType),
    StructField("event_type", StringType),
    StructField("product_id", LongType),
    StructField("country", StringType),
    StructField("amount", DoubleType),
    StructField("currency", StringType),
    StructField("quantity", IntegerType),
    StructField("event_ts", LongType),
    StructField("ingest_ts", LongType),
    StructField("payload", StringType)
  ))

  def main(args: Array[String]): Unit = {
    val a = args.sliding(2, 2).collect { case Array(k, v) => k.stripPrefix("--") -> v }.toMap
    val bootstrap = sys.env.getOrElse("KAFKA_BOOTSTRAP", "kafka:9092")
    val topic = sys.env.getOrElse("SOURCE_TOPIC", "events")
    val db = sys.env.getOrElse("ICEBERG_DB", "streaming")
    val table = sys.env.getOrElse("ICEBERG_TABLE", "events") + "_spark"
    val mode = a.getOrElse("distribution-mode", sys.env.getOrElse("WRITE_DISTRIBUTION_MODE", "hash"))
    val upsert = a.getOrElse("upsert", "false").toBoolean
    val triggerSec = a.getOrElse("trigger", sys.env.getOrElse("CHECKPOINT_INTERVAL_MS", "30000").toInt / 1000 max 1).toString
    // partitioning: "bucket" = bucket(N,user_id) | "time" = hours(event_time)
    val partitioning = a.getOrElse("partitioning", sys.env.getOrElse("PARTITIONING", "bucket"))
    // bucket count = concurrent write streams (see Flink job note). Default 48 to
    // match Flink so the comparison holds the write-parallelism knob equal.
    val buckets = a.getOrElse("buckets", sys.env.getOrElse("BUCKETS", "48"))
    val formatVersion = a.getOrElse("format-version", sys.env.getOrElse("TABLE_FORMAT_VERSION", "2"))
    // Upsert engine mode: "merge-on-read" (deletion vectors / delete files, cheap
    // writes + read-time merge) vs "copy-on-write" (rewrite touched data files on
    // each MERGE, heavier writes + clean reads). Set on the delete/update/merge
    // table properties so the MERGE in foreachBatch honours it. Lets us compare BOTH
    // Spark patterns against Flink's DV upsert. Ignored for append.
    val mergeMode = a.getOrElse("merge-mode", sys.env.getOrElse("WRITE_MERGE_MODE", "merge-on-read"))
    // "time" partitions on epoch-minute (identity) so a 5-min run spreads across
    // ~5 partitions (Iceberg has no minute() transform); "bucket" = cardinality.
    val partitionClause = if (partitioning == "time") "event_minute" else s"bucket($buckets, user_id)"

    // Catalog config source depends on environment:
    //  * EKS (Spark operator): ALL spark.sql.catalog.demo.* come from the manifest
    //    sparkConf (baked into spark-defaults.conf). We must NOT re-set them here —
    //    the operator pre-creates the SparkSession, so builder .config() calls are
    //    DROPPED by getOrCreate(), and Iceberg's GlueCatalog then initializes with
    //    empty props → wrong S3 region → 301. Detected via CONFIG_FROM_MANIFEST=1.
    //  * Local: we set them programmatically (REST catalog on MinIO).
    val catalogType = sys.env.getOrElse("CATALOG_TYPE", "rest")
    val warehouse = s"s3://${sys.env.getOrElse("WAREHOUSE_BUCKET", "warehouse")}/"
    val configFromManifest = sys.env.getOrElse("CONFIG_FROM_MANIFEST", "0") == "1"
    val checkpointBase = sys.env.getOrElse("CHECKPOINT_BASE", "file:///tmp/spark-ckpt")

    val builder = SparkSession.builder()
      .appName(s"spark-structured-streaming-iceberg[$mode${if (upsert) ",upsert" else ""}]")
    if (!configFromManifest) {
      // Local path: set catalog config in code (manifest/spark-defaults not used).
      builder
        .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
        .config("spark.sql.catalog.demo", "org.apache.iceberg.spark.SparkCatalog")
        .config("spark.sql.catalog.demo.type", catalogType)
        .config("spark.sql.catalog.demo.warehouse", warehouse)
        .config("spark.sql.catalog.demo.io-impl", "org.apache.iceberg.aws.s3.S3FileIO")
      if (catalogType == "rest") {
        builder
          .config("spark.sql.catalog.demo.uri", sys.env.getOrElse("CATALOG_URI", "http://iceberg-rest:8181"))
          .config("spark.sql.catalog.demo.s3.endpoint", sys.env.getOrElse("S3_ENDPOINT", "http://minio:9000"))
          .config("spark.sql.catalog.demo.s3.path-style-access", "true")
      }
    }
    val spark = builder.getOrCreate()
    import spark.implicits._

    // ── Storage-Partitioned Join (SPJ) for the upsert MERGE ──────────────────
    // The target table is bucket(N, user_id) and the MERGE key is user_id, so the
    // target is ALREADY laid out by the join key. With SPJ on, Spark can co-partition
    // the source micro-batch to the target's bucketing and skip the full shuffle-to-
    // target on every batch. These are session-level SQL configs (safe to set in
    // code even under CONFIG_FROM_MANIFEST). Toggle with SPJ=1 to measure the delta.
    //   NOTE: our internal "autospj" auto-invokes this at the engine layer (source
    //   auto-partitioned to the target spec). Here we enable the OSS-available knobs
    //   + explicitly pre-cluster the source, which is the reproducible equivalent.
    if (sys.env.getOrElse("SPJ", "0") == "1") {
      spark.conf.set("spark.sql.sources.v2.bucketing.enabled", "true")
      spark.conf.set("spark.sql.iceberg.planning.preserve-data-grouping", "true")
      spark.conf.set("spark.sql.sources.v2.bucketing.pushPartValues.enabled", "true")
      spark.conf.set("spark.sql.requireAllClusterKeysForCoPartition", "false")
      spark.conf.set("spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled", "true")
    }
    val spjEnabled = sys.env.getOrElse("SPJ", "0") == "1"

    val fqTable = s"demo.$db.$table"
    spark.sql(s"CREATE DATABASE IF NOT EXISTS demo.$db")
    // Clean per-run state: drop+recreate the table so each run measures only its
    // own output (RECREATE_TABLE=1). PURGE removes the data files too, so the
    // engine self-cleans without any gated AWS CLI call.
    if (sys.env.getOrElse("RECREATE_TABLE", "0") == "1") {
      // Plain DROP (no PURGE): PURGE triggers a DataSourceV2Relation code path that
      // NoSuchMethodErrors on this iceberg-4.0 / Spark 4.1.2 combo. Metadata-only
      // drop is enough — per-run measurement reads the current snapshot, so any
      // orphaned data files from a prior run aren't counted.
      spark.sql(s"DROP TABLE IF EXISTS $fqTable")
    }
    // For upsert, set delete/update/merge modes to the chosen engine (MoR vs CoW).
    // MoR needs format-version >= 2 (v3 for deletion vectors); the run passes FMT=3.
    val mergeProps = if (upsert)
      s""",
         |  'write.delete.mode'='$mergeMode',
         |  'write.update.mode'='$mergeMode',
         |  'write.merge.mode'='$mergeMode'""".stripMargin
    else ""
    spark.sql(
      s"""CREATE TABLE IF NOT EXISTS $fqTable (
         |  event_id string, user_id bigint, event_type string, product_id bigint,
         |  country string, amount double, currency string, quantity int,
         |  event_ts bigint, ingest_ts bigint, payload string,
         |  event_time timestamp, event_minute bigint)
         |USING iceberg
         |PARTITIONED BY ($partitionClause)
         |TBLPROPERTIES (
         |  'format-version'='$formatVersion',
         |  'write.distribution-mode'='$mode'$mergeProps
         |)""".stripMargin)

    // maxOffsetsPerTrigger is UNSET by default so Spark drains as fast as it can
    // each micro-batch — matching Flink's uncapped continuous read for a fair
    // overload comparison. Set MAX_OFFSETS_PER_TRIGGER only to deliberately throttle.
    val maxOffsets = sys.env.getOrElse("MAX_OFFSETS_PER_TRIGGER", "")
    val rawReader = spark.readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", bootstrap)
      .option("subscribe", topic)
      .option("startingOffsets", a.getOrElse("starting-offsets", sys.env.getOrElse("STARTING_OFFSETS", "latest")))
      // aggressive Kafka retention can purge offsets a stale checkpoint expects;
      // don't fail the query on that.
      .option("failOnDataLoss", "false")
    val raw = (if (maxOffsets.nonEmpty) rawReader.option("maxOffsetsPerTrigger", maxOffsets) else rawReader)
      .load()

    val parsed: DataFrame = raw
      .select(from_json(col("value").cast("string"), schema).as("e"))
      .select("e.*")
      // derive time columns: a real timestamp + an epoch-minute bucket for
      // minute-granular identity partitioning.
      .withColumn("event_time", (col("event_ts") / 1000).cast("timestamp"))
      .withColumn("event_minute", (col("event_ts") / 60000).cast("long"))

    val query = if (upsert) {
      // Upsert path: MERGE INTO on each micro-batch keyed by user_id — the same
      // equality key Flink's .upsert() uses, so the two are comparable. Iceberg
      // MERGE errors if >1 source row matches one target row, so we dedup the
      // batch to the latest event per user_id (highest ingest_ts) first.
      // In-submit compaction modes — modeled on AWS's iceberg-streaming-examples
      // (SparkCustomIcebergIngest: compaction=none|inline|scheduled). Spark has NO
      // in-query maintenance operator like Flink, so the options are:
      //   none      — no in-submit compaction (baseline; use a separate scheduled job,
      //               which AWS calls the RECOMMENDED production baseline — our compact_spark.sh)
      //   inline    — run compaction inside foreachBatch every N batches. Simple, but
      //               SYNCHRONOUS → lengthens those batches / stalls ingest.
      //   scheduled — background ScheduledExecutorService fires compaction on a timer,
      //               concurrent with ingest on the SAME submit's executors (still
      //               competes with the writer, but doesn't block the batch).
      // Every maintenance call is guarded (a failed compaction must NEVER kill ingest).
      val compactMode = sys.env.getOrElse("SPARK_COMPACTION", "none")   // none|inline|scheduled
      val maintEvery = sys.env.getOrElse("SPARK_INLINE_EVERY", "10").toInt
      val maintExpire = sys.env.getOrElse("SPARK_MAINT_EXPIRE", "1") == "1"
      val maintGate = new Semaphore(1)  // at most one compaction in flight (drop overlaps)
      def runMaintenance(s: SparkSession, why: String): Unit = {
        if (!maintGate.tryAcquire()) return  // previous compaction still running — skip
        try {
          s.sql(s"CALL demo.system.rewrite_data_files(table => '$db.$table', options => map('min-input-files','2'))")
          s.sql(s"CALL demo.system.rewrite_position_delete_files(table => '$db.$table')")
          if (maintExpire) s.sql(s"CALL demo.system.expire_snapshots(table => '$db.$table', retain_last => 5)")
        } catch {
          case e: Throwable => System.err.println(s"[maint:$why] failed (ingest continues): ${e.getMessage}")
        } finally maintGate.release()
      }
      // scheduled mode: background timer thread on the driver, concurrent with ingest.
      val scheduler: java.util.concurrent.ScheduledExecutorService =
        if (compactMode == "scheduled") {
          val sc = Executors.newSingleThreadScheduledExecutor((r: Runnable) => {
            val th = new Thread(r, "sfi-compaction"); th.setDaemon(true); th
          })
          val everySec = sys.env.getOrElse("SPARK_SCHEDULED_EVERY_SEC", "120").toLong
          sc.scheduleAtFixedRate(() => runMaintenance(spark, "scheduled"), everySec, everySec,
            java.util.concurrent.TimeUnit.SECONDS)
          sc
        } else null
      parsed.writeStream
        .foreachBatch { (batch: DataFrame, batchId: Long) =>
          val deduped = batch
            .withColumn("_rn", row_number().over(
              Window.partitionBy("user_id").orderBy(col("ingest_ts").desc)))
            .filter(col("_rn") === 1).drop("_rn")
          // SPJ: cluster the source by the join key so the planner can co-partition
          // it with the target's bucket(N, user_id) layout and skip the shuffle-to-
          // target inside MERGE. We repartition by user_id into `buckets` partitions
          // (matching the target bucket count); with preserve-data-grouping +
          // v2.bucketing on, Spark's KeyGroupedPartitioning aligns this with the
          // target's bucket transform. (NOTE: `bucket()` is an Iceberg DDL transform,
          // not a DataFrame function — hash-repartitioning by the key is the correct
          // OSS way to feed SPJ here.) Without SPJ this branch is skipped.
          val src =
            if (spjEnabled && partitioning != "time")
              deduped.repartition(buckets.toInt, col("user_id"))
            else deduped
          src.createOrReplaceTempView("updates")
          batch.sparkSession.sql(
            s"""MERGE INTO $fqTable t
               |USING (SELECT * FROM updates) s
               |ON t.user_id = s.user_id
               |WHEN MATCHED THEN UPDATE SET *
               |WHEN NOT MATCHED THEN INSERT *""".stripMargin)

          // inline mode: compact synchronously inside the batch every N batches
          // (blocks the next batch — the AWS "simple but lengthens batches" pattern).
          if (compactMode == "inline" && batchId > 0 && batchId % maintEvery == 0) {
            runMaintenance(batch.sparkSession, s"inline@batch$batchId")
          }
          () // foreachBatch requires a Unit-returning closure
        }
        .option("checkpointLocation", s"${checkpointBase}/$table")
        .trigger(Trigger.ProcessingTime(s"$triggerSec seconds"))
        .start()
    } else {
      // Append path: native DataSourceV2 streaming writer.
      parsed.writeStream
        .format("iceberg")
        .outputMode("append")
        .option("checkpointLocation", s"${checkpointBase}/$table")
        .option("fanout-enabled", "true")
        .trigger(Trigger.ProcessingTime(s"$triggerSec seconds"))
        .toTable(fqTable)
    }

    query.awaitTermination()
  }
}
